src/Algebra/Module/Morphism/Linear/Properties.agda

```diff
@@ -0,0 +1,163 @@
+
+open import Axiom.DoubleNegationElimination
+open import Data.List
+open import Data.Product
+  using (Σ; _,_; ∃; Σ-syntax; ∃-syntax; _×_)
+open import Relation.Nullary          using (¬_)
+open import Relation.Nullary.Negation using (contraposition)
+
+open MorphismStructures.ModuleMorphisms modA modB
+module A = Module modA
+open A using () renaming (Carrierᴹ to A)
+-- module A = ModuleProperties A.leftSemimodule
+-- open ModuleProperties A.leftSemimodule
+--   using () renaming
+--   ( non-zeroʳ to non-zeroʳᴬ
+--   ; non-zeroˡ to non-zeroˡᴬ
+--   )
+module B = Module modB
+open B using () renaming (Carrierᴹ to B)
+module B2 = ModuleProperties B.leftSemimodule
+-- open ModuleProperties B.leftSemimodule
+--   using () renaming
+--   ( non-zeroʳ to non-zeroʳᴮ
+--   ; non-zeroˡ to non-zeroˡᴮ
+--   )
+open CommutativeRing ring
+  using (_≈_; _*_; _+_; 0#) renaming (Carrier to S)
+
+module _
+  {f : A → B}
+  (isModuleHomomorphism : IsModuleHomomorphism f)
+  where
+
+  open IsModuleHomomorphism isModuleHomomorphism
+
+  -- f(x) ≈ 0 iff x ≈ 0, for linear non-trivial f
+  f𝟘≈𝟘 : f A.0ᴹ B.≈ᴹ B.0ᴹ
+  f𝟘≈𝟘 = begin
+    f A.0ᴹ           ≈⟨ ⟦⟧-cong (A.≈ᴹ-sym (A.*ₗ-zeroˡ A.0ᴹ)) ⟩
+    f (0# A.*ₗ A.0ᴹ) ≈⟨ *ₗ-homo 0# A.0ᴹ ⟩
+    0# B.*ₗ f A.0ᴹ   ≈⟨ B.*ₗ-zeroˡ (f A.0ᴹ) ⟩
+    B.0ᴹ             ∎
+    where open Reasoning B.≈ᴹ-setoid
+
+  x≈𝟘⇒fx≈𝟘 : ∀ {x} → x A.≈ᴹ A.0ᴹ → f x B.≈ᴹ B.0ᴹ
+  x≈𝟘⇒fx≈𝟘 {x} x≈𝟘 = begin
+    f x    ≈⟨ ⟦⟧-cong x≈𝟘 ⟩
+    f A.0ᴹ ≈⟨ f𝟘≈𝟘 ⟩
+    B.0ᴹ   ∎
+    where open Reasoning B.≈ᴹ-setoid
+
+  fx≉𝟘⇒x≉𝟘 : ∀ {x} → f x B.≉ᴹ B.0ᴹ → x A.≉ᴹ A.0ᴹ
+  fx≉𝟘⇒x≉𝟘 = contraposition x≈𝟘⇒fx≈𝟘
+
+  -- Zero is a unique output of linear map ≉ `const 𝟘`.
+  zero-unique : ∀ {x} →
+    Σ[ (s , y) ∈ S × A ] ((s A.*ₗ x A.≈ᴹ y) × (f y B.≉ᴹ B.0ᴹ)) →
+    x A.≉ᴹ A.0ᴹ → f x B.≉ᴹ B.0ᴹ
+  zero-unique {x} ((s , y) , (s·x≈y , fy≉𝟘)) x≉𝟘 =
+    B2.non-zeroʳ s·fx≉𝟘
+    where
+    open Reasoning B.≈ᴹ-setoid
+    y≉𝟘     : y A.≉ᴹ A.0ᴹ
+    y≉𝟘     = fx≉𝟘⇒x≉𝟘 fy≉𝟘
+    fs·x≈fy : f (s A.*ₗ x) B.≈ᴹ f y
+    fs·x≈fy = ⟦⟧-cong s·x≈y
+    s·fx≈fy : s B.*ₗ f x B.≈ᴹ f y
+    s·fx≈fy = begin
+      s B.*ₗ f x   ≈⟨ B.≈ᴹ-sym (*ₗ-homo s x) ⟩
+      f (s A.*ₗ x) ≈⟨ fs·x≈fy ⟩
+      f y          ∎
+    s·fx≉𝟘  : (s B.*ₗ f x) B.≉ᴹ B.0ᴹ
+    s·fx≉𝟘 s·fx≈𝟘 =
+      fy≉𝟘 ( begin
+             f y          ≈⟨ ⟦⟧-cong (A.≈ᴹ-sym s·x≈y) ⟩
+             f (s A.*ₗ x) ≈⟨ *ₗ-homo s x ⟩
+             s B.*ₗ f x   ≈⟨ s·fx≈𝟘 ⟩
+             B.0ᴹ         ∎
+           )
+
+  -- f is odd (i.e. - f (-x) ≈ - (f x)).
+  fx+f[-x]≈𝟘 : ∀ {x} → f x B.+ᴹ f (A.-ᴹ x) B.≈ᴹ B.0ᴹ
+  fx+f[-x]≈𝟘 {x} = begin
+    f x B.+ᴹ f (A.-ᴹ x) ≈⟨ B.≈ᴹ-sym (+ᴹ-homo x (A.-ᴹ x)) ⟩
+    f (x A.+ᴹ (A.-ᴹ x)) ≈⟨ ⟦⟧-cong (A.-ᴹ‿inverseʳ x) ⟩
+    f A.0ᴹ              ≈⟨ f𝟘≈𝟘 ⟩
+    B.0ᴹ                ∎
+    where open Reasoning B.≈ᴹ-setoid
+
+  f[-x]≈-fx : ∀ {x} → f (A.-ᴹ x) B.≈ᴹ B.-ᴹ f x
+  f[-x]≈-fx {x} = B.uniqueʳ‿-ᴹ (f x) (f (A.-ᴹ x)) fx+f[-x]≈𝟘
+
+  module _ {dne : DoubleNegationElimination _} where
```

Comment:
Hmm assuming double negation elimination (for every type!) is very strong here! First of all is definitely needed for all of these? (just checking!)

Secondly can we get away with assuming the (weaker?) statement of decidability of `A.≉ᴹ`?

Reply:
No, I only need: `¬x≉0⇒x≈0`. I just wasn't sure how else to get that.

You think I can get it, via decidability?
Can you give me a bit more of a hint?

Reply:
I would put that locally as an assumption only where you need it.  Yes, this is derivable from decidability.

Reply:
> I would put that locally as an assumption only where you need it.

Yeah, that makes sense since I only use it for one lemma.

**==> Requested change arriving in next push.**